---
title: "ggplot2"
author: "kat"
date: "October 14, 2015"
output: html_document
---

##1. Subsetting Data

Loading in data, and manipulating a few columns
```{r loading in data, echo=TRUE}
d <- read.csv("Dataset_S1.txt", sep= ',', header=TRUE) #reading in data and calling it d
head(d) #checking data to see if imported correctly
colnames(d) [12] <- "percent.GC" #renamed column 12 percent.GC 
d$diversity <- d$Pi/(10*1000) # rescale  diversity estimates from last class
d$cent <- d$start >= 25800000 & d$end <= 29700000 #identify each row as in being in the centromere or not in the centromere 
```

Which values of the vector conforms to our specifications for SNPs
```{r SNPs, echo=TRUE}
d$total.SNPs >= 85 #number of SNPs that are greater than or equal to 85 (logical)
greaterequal85.SNPs <- d[d$total.SNPs >=85,] #this extracts rows with greater than or equal to 85 SNPs, this gave us only 3 rows 
```


###Subsetting on 2 columns of data
```{r subsetting by 2 columns, echo=TRUE}
d[d$Pi >16 & d$percent.GC >80, c("start", "end", "depth", "Pi")] #pulling out specified data from these particular columns start, end, etc
#if you wanted to re-order columns in output, reorder the items in cbind
```

###summarizing data after subsetting
```{r summarizing subset data, echo=TRUE}
summary(d$depth[d$percent.GC >=80])
summary(d$depth[d$percent.GC <80])
```

###subset using subset function
```{r subset function, echo=TRUE}
subset(d, Pi >16 & percent.GC >80, c(start, end, Pi, percent.GC, depth))
```


##2. Exploring data visually with ggplot2

```{r loading ggplot, echo=FALSE}
#install.packages("ggplot2") #remove initial # if need to install
#library(ggplot2) #calling ggplot2
```

###making plots
```{r inital diversity plot, echo=TRUE}
d$position <- (d$end + d$start) / 2 #making a mid-point (new column)
ggplot(d) + geom_point(aes(x=position, y=diversity)) #basic plot, + adds a layer, here we're plotting a geometric point
ggplot(d) + geom_point(aes(x=position, y=diversity, color=cent)) #adding color to the true and false
```


####reducing overplotting
```{r reduce overplotting, echo=TRUE}
ggplot(d) + geom_density(aes(x=diversity), fill="black") #making density plot of diversity values
ggplot(d) + geom_density(aes(x=diversity, fill=cent), alpha=0.4) #the areas near centromere have slightly more diversity than areas away from the centromere alpha=transparency value
```

####reducing resolution of data by binning, taking continuous numeric values and puts them in discrete number of bins

```{r bin, echo=TRUE}
d$GC.binned <- cut(d$percent.GC, 5) #using cut to bin the percent GC into 5 bins
ggplot(d) + geom_bar(aes(x=GC.binned)) #r putting into own bins, default 30
ggplot(d) + geom_bar(aes(x=percent.GC)) #plotting percent GC that we put in bins
ggplot(d) + geom_density(aes(x=depth, linetype=GC.binned), alpha=0.5) #this shows us the lowest and highest GC content windows have lower depth
```

##3. Merging and Combining Data

loading in our data files
```{r loading in data for merging, echo=TRUE}
mtfs <- read.delim("motif_recombrates.txt", header=TRUE) #read in motif recombination rates data
rpts <- read.delim("motif_repeats.txt", header=TRUE) #read in motif repeats
head(rpts,3) #looking at the first 3 lines of rpts

mtfs$pos <-paste(mtfs$chr, mtfs$motif_start, sep="-") #make a column of specific positions per chromosome in the metfs dataframe
rpts$pos <- paste(rpts$chr, rpts$motif_start, sep="-") #make a column of a specific positions per chromosome in rpts dataframe
recm <-merge (mtfs, rpts, by.x="pos", by.y="pos") #merge making a new dataframe 
head(recm)
nrow(recm) #when merging, want to make sure that data hasn't been doubled. want less data than when you started with 
```

Left outer join to inclue all records from first file (the x position)
```{r outer join, echo=TRUE}
recm2 <- merge(mtfs, rpts, by.x="pos", by.y="pos", all.x=TRUE) #the merge statement with a left outer join all.y would be a right outer join
head(recm2) #look at the data that were merged
nrow(recm2) #how many rows in new dataframe?
```


##4. Working with Split-Apply-Combind Pattern

Split data according to depth and GC.binned
```{r split data, echo=TRUE}
d_split <- split(d$depth, d$GC.binned)
str(d_split)
lapply(d_split, mean) #uses the lapply function to calculate the mean of each group

```

combine our split data into a single vector
```{r combine data back together, echo=TRUE}
unlist(lapply(d_split,mean)) #uses unlist to make a vector out of the list
sapply(d_split, mean) #equivalent to above 2 commands
```

Want more values than just mean
```{r summary, echo=TRUE}
dpth_sum <- lapply(split(d$depth, d$GC.binned), summary)
dpth_sum
```

Then put all of these values back together in the dataframe. However if you have many rows or columns to bind together into a dataframe
```{r rbind, echo=TRUE} 
rbind(dpth_sum[[1]], dpth_sum[[2]]) #uses rbind to bind together 2 rows #rbind = rowbind
```

do call function
```{r do.call, echo=TRUE}
do.call(rbind, lapply(split(d$depth, d$GC.binned), summary)) #taking all of the rows and binds them together
```

Aggregate function
```{r aggregate, echo=TRUE}
aggregate(d$depth, list(gc=d$GC.binned), mean)
```


